### ISSUE TITLE
wdio.conf.js debug not working in 4.9.9

### ISSUE THREAD

**Issue by mkj28** [webdriverio/webdriverio#2467]
## The problem

Looks like https://github.com/webdriverio/webdriverio/commit/f878fbaa78f8d083e6fe0dacad5f770166b7ff96 broke the `debug` option in wdio.conf.js

## Environment

* WebdriverIO version:4.9.9
* Node.js version:N/A
* [Standalone mode or wdio testrunner](http://webdriver.io/guide/getstarted/modes.html):N/A
* if wdio testrunner, running synchronous or asynchronous tests:N/A
* Additional wdio packages used (if applicable):N/A

## Details

Debugger no longer starting in 4.9.9 if `debug: true`

## Link to Selenium/WebdriverIO logs

N/A

## Code To Reproduce Issue [ Good To Have ]

N/A

**Comment by mkj28** [comment]
Maybe I am missing something but running `wdio --debug ...` does not start the debugger for me either...

**Comment by christian-bromann** [comment]
What node version are you running? (Please always submit that info!)

**Comment by mkj28** [comment]
> What node version are you running?

Ugh, sorry, did not think it's relevant: v6.9.2

**Comment by mkj28** [comment]
> What node version are you running?

same on node 8.9.1 - debugger not starting when `debug: true`

**Comment by christian-bromann** [comment]
@codiophile can you have a look at this? Looks like I need to revert your change.

**Comment by codiophile** [comment]
@mkj28 The way it works now is that the children inherit the debug flags from the parent. We should update the documentation or discuss why this isn't preferable. The wdio debug flag doesn't do anything anymore. Now the only flag that matters is the debug/inspect flag of the node process. Please let me know if this is a problem for you.

**Comment by christian-bromann** [comment]
> The wdio debug flag doesn't do anything anymore.

This should definitely be updated in the docs

**Comment by codiophile** [comment]
@christian-bromann I'll update the docs. Do we have a way of marking deprecated configuration options as deprecated?

**Comment by christian-bromann** [comment]
> Do we have a way of marking deprecated configuration options as deprecated?

Not really, just output a message saying that this is deprecated like we did with the commands

**Comment by codiophile** [comment]
Out of curiosity, did `wdio --debug` ever work? Also, if you run `wdio` using the `wdio` command, how do you start node with debugger flags?

**Comment by mkj28** [comment]
> Also, if you run wdio using the wdio command, how do you start node with debugger flags?

same question... how do you do that? Before this change - you could easily debug webdriver.io from VSCode, not even sure if that is still doable with this change? I guess you would have to run node in debug and then attach VSCode debugger? Sounds complex to me...

**Comment by codiophile** [comment]
@mkj28 I think it should be possible to add `execArgv: ['--inspect']` (or `execArgv: ['--debug']` depending on node version) to the config as a workaround.

**Comment by mkj28** [comment]
> should be possible to add execArgv: ['--inspect'] (or execArgv: ['--debug'] depending on node version) to the config as a workaround.

thanks, so for me this worked around it:
`execArgv: process.env.DEBUG ? ['--debug=5859'] : [],`
(since my launch.json was already setting DEBUG=true and was picked up as `debug: process.env.DEBUG` in wdio.conf.js

**Comment by christian-bromann** [comment]
@codiophile what solution to this would you propose?

**Comment by codiophile** [comment]
> @codiophile what solution to this would you propose?

@christian-bromann That's a good question. We group our specs based on environment configuration, so that we don't need to reconfigure the environment too often. Because of this we have our own cli that uses the wdio launcher internally. So we don't actually use the `wdio` command.

For us it is very convenient that the child process inherits the debug flags from the parent, since this just works automagically when debugging in WebStorm. I understand that we need a solution that also works for people using the `wdio` command, but unfortunately I don't have a project I can test this with. 

    execArgv: process.env.DEBUG ? ['--debug=5859'] : [],

The above seems like a decent workaround for now. We leave it to the user to decide whether to use debug or inspect. I suppose we could make the syntax nicer, by doing something like:

    debug: 'debug' // For --debug
    debug: 'inspect' // For --inspect

If we stick with `debug: true`, we would need to choose `--debug` or `--inspect` based on the node version. The tricky part here is to guess which method a user of a node version that supports both wants to use. I'd rather leave it to the user to decide.

One thing I haven't tried, but think would work is to use `$NODE_DEBUG_OPTION` in your npm script, but I think that might be a WebStorm specific feature.

Preferably I'd like to find a solution that just works and is compatible with how IDE's would launch `wdio` in debug mode. I'd prefer if we didn't have to edit the config file and debugging just works out of the box, which is the case for us today. I'll probably have to experiment a bit to find out how to do that with the `wdio` command, but here's what I think would work already today:

    "scripts" : {
        "test": "node $NODE_DEBUG_OPTION ./node_modules/.bin/wdio"
    }

To the best of my knowledge that should work automagically if you're using WebStorm and if you're using something else I suppose it wouldn't be too hard to manually configure it to set `$NODE_DEBUG_OPTION` to `--inspect` or whatever you need it to be. I'm not sure we need to change anything but the documentation for this. We could bring back the `debug` option, but I'm not sure that's desirable. I don't think the wdio config is the desirable place to configure debugging; I think `package.json` is a much more sensible place as it has to do with how webdriverio is launched.

To finally answer your question, I would propose that we verify that the above npm script works as expected and add that to the documentation. We could also document the `execArgv` workaround for people that don't want to use npm script and prefer to run their global `wdio`, but I would strongly recommend against that. I suppose we could integrate this into `wdio` once we implement #2421, but I don't think most people interested in debugging would need that much handholding.

Let me know your thoughts. I can start working on the required pull requests, as soon as I know what we want to do.

**Comment by codiophile** [comment]
After thinking about this a bit more, I think a nice solution would be to implement support for  
`--debug` and `--inspect` for the `wdio` command. It shouldn't be too much effort.

**Comment by tarun** [comment]
Another thing I noticed is - if there is a command line argument `--expose_debug_as=v8debug` (which IntelliJ automatically adds for node 6) - the wdio argument parser errors out

```
/xyz/node_modules/webdriverio/build/lib/cli.js:438
                throw e;
                ^

TypeError: Invalid attempt to destructure non-iterable instance
    at /xyz/node_modules/babel-runtime/helpers/slicedToArray.js:48:13
    at Launcher.startInstance (/xyz/node_modules/webdriverio/build/lib/launcher.js:563:74)
    at Launcher.runSpecs (/xyz/node_modules/webdriverio/build/lib/launcher.js:504:22)
    at /xyz/node_modules/webdriverio/build/lib/launcher.js:354:47
    at Promise.F (/xyz/node_modules/babel-runtime/node_modules/core-js/library/modules/_export.js:35:28)
    at Launcher._callee$ (/xyz/node_modules/webdriverio/build/lib/launcher.js:348:40)
    at tryCatch (/xyz/node_modules/regenerator-runtime/runtime.js:65:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/xyz/node_modules/regenerator-runtime/runtime.js:299:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/xyz/node_modules/regenerator-runtime/runtime.js:117:21)
    at step (/xyz/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)

```

in `launcher.js` __process$execArgv$i$_ evaluates to _null_
```
for (var i in process.execArgv) {
    var _process$execArgv$i$m = process.execArgv[i].match('--(debug|inspect)(?:-brk)?(?:=(.*):)?'),
        _process$execArgv$i$m2 = (0, _slicedToArray3.default)(_process$execArgv$i$m, 3),
```

**Comment by codiophile** [comment]
@tarun I would say that's a separate issue, but thanks for reporting it. I'll try to make a PR for that today or tomorrow.

**Comment by codiophile** [comment]
@tarun Please check if #2490 solves the problem for you.

**Comment by christian-bromann** [comment]
Fixed by #2492
